Fix the GitLab cloud provider, add unit tests

qmfrederik opened the pull request:
This improves how the GitLab cloud provider handles refs and tags (although some level of guesswork is involved, as Gitlab doesn't provide raw access to the full ref).

Also adds unit tests. This requires mocking the `Environment.GetEnvironmentVariable` call, so there now is a `CloudBuildEnvironment` class with just a single `GetEnvironmentVariable` method which can be overriden for unit tests.

A bunch of classes are no longer `static` so we can use constructor DI on them.

Fixes #335

AArnott commented on src/NerdBank.GitVersioning/CloudBuildServices/GitLab.cs:
```diff
@@ -15,15 +14,26 @@
+
+        public string BuildingBranch =>
+            this.buildEnvironment.GetEnvironmentVariable("CI_COMMIT_TAG") == null ? 
+                $"/refs/heads/{this.buildEnvironment.GetEnvironmentVariable("CI_COMMIT_REF_NAME")}" : null;
```
There should be no leading `/` in the return value

AArnott commented on src/NerdBank.GitVersioning.Tests/CloudBuildServices/GitLabTests.cs:
```diff
@@ -0,0 +1,53 @@
+
+
+            Assert.NotNull(cloudBuild.Active);
+            Assert.Equal("/refs/heads/master", cloudBuild.Active.BuildingBranch);
```
Update this assertion as well to *not* expect a leading slash.

AArnott commented on src/NerdBank.GitVersioning/CloudBuildEnvironment.cs:
```diff
@@ -0,0 +1,30 @@
+        /// Gets the default instance of the <see cref="CloudBuildEnvironment"/> class.
+        /// </summary>
+        public static CloudBuildEnvironment Instance
+        { get; set; } = new CloudBuildEnvironment();
```
Move this up to the end of the previous line, please.

AArnott commented on src/NerdBank.GitVersioning/CloudBuildEnvironment.cs:
```diff
@@ -0,0 +1,30 @@
+    /// <summary>
+    /// Provides access to the current system's environment variables.
+    /// </summary>
+    public class CloudBuildEnvironment
```
This class should be internal. It's not part of the public contract I wish to maintain with folks who reference this library.

qmfrederik commented:
@AArnott Sure, I updated the branch to use the existing pattern, and removed the leading /.

I had to update the constructor of the `BuildIntegrationTests` to not directly use any MSBuild-related types; otherwise the unit tests would fail with this error:

```
System.InvalidOperationException : Microsoft.Build.Locator.MSBuildLocator.RegisterInstance was called, but MSBuild assemblies were already loaded.
Ensure that RegisterInstance is called before any method that directly references types in the Microsoft.Build namespace has been called.
Loaded MSBuild assemblies: Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Microsoft.Build.Framework, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Stack Trace:
   at Microsoft.Build.Locator.MSBuildLocator.RegisterMSBuildPath(String msbuildPath)
   at Microsoft.Build.Locator.MSBuildLocator.RegisterInstance(VisualStudioInstance instance)
   at Microsoft.Build.Locator.MSBuildLocator.RegisterDefaults()
   at MSBuildExtensions.LoadMSBuild()
   at BuildIntegrationTests..ctor(ITestOutputHelper logger)
```

Let me know if this works for you.

AArnott commented:
> I updated the branch to use the existing pattern, and removed the leading /.

Thanks

> I had to update the constructor of the BuildIntegrationTests to not directly use any MSBuild-related types; otherwise the unit tests would fail with this error:

I already fixed it [here](https://github.com/AArnott/Nerdbank.GitVersioning/pull/339/files#diff-671f15e0b30fa54620eef736ecea8be5). If you revert your own fix for it and merge from my `v2.3` branch, you'll get the fix too.

Note that over the weekend I bumped `master` to `v3.0-beta`, so I'll retarget you PR to the `v2.3` branch since that's effectively what you were targeting before.

AArnott commented:
Never mind. I can't retarget it because you've already rebased onto the new master branch. May I recommend you rebase your commits onto v2.3 and change the target branch of this PR yourself when ready?

AArnott commented on src/NerdBank.GitVersioning/CloudBuildServices/GitLab.cs:
```diff
@@ -15,11 +13,15 @@
+                $"refs/heads/{Environment.GetEnvironmentVariable("CI_COMMIT_REF_NAME")}" : null;
 
-        public string BuildingRef => Environment.GetEnvironmentVariable("CI_COMMIT_TAG");
+        public string BuildingRef => this.BuildingBranch != null ? this.BuildingBranch : this.BuildingTag;
```
```suggestion
        public string BuildingRef => this.BuildingBranch ?? this.BuildingTag;
```

qmfrederik commented:
>> I had to update the constructor of the BuildIntegrationTests to not directly use any MSBuild-related types; otherwise the unit tests would fail with this error:

> I already fixed it here. If you revert your own fix for it and merge from my v2.3 branch, you'll get the fix too.

That fix doesn't seem to be sufficient, and I had to move all code which references any MSBuild-related type out of the constructor.

> May I recommend you rebase your commits onto v2.3 and change the target branch of this PR yourself when ready?

Sure, done.

AArnott commented:
> That fix doesn't seem to be sufficient, and I had to move all code which references any MSBuild-related type out of the constructor.

Strange. But oh well. Thanks for contributing!
